### DISCUSSION TITLE
Trying compiler

### DISCUSSION THREAD

**Discussion by dac2020** [arendst/Tasmota#16384] [Q&A]
I'm trying my hand at compiling.

The info says to modify user_config_override.h for your needs but I don't see that many options so wondering if there is some documentation I need to look at to learn more.

For example, I don't see any option for which chip to build for but seems to be for esp8266. 
I also see no options for TLS but I changed the port to 8883 so wondering if this means the compiler will know to add TLS?

I compiled and got an error which says to report this. Most likely, it's just me missing something.
I used root as the user since there is no one else on this system and I didn't want to mess with sudo.
Maybe that is my problem however.

The OS is centos 8.
The error doesn't appear to be permissions or directories or files missing so I'm not really sure what I should be looking for to solve this.

'''
***snip***
Platform Manager: Installing https://github.com/tasmota/platform-espressif32/releases/download/v2.0.4.1/platform-espressif32-2.0.4.1.zip
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Platform Manager: espressif32@2.0.4 has been installed!
Tool Manager: Installing https://github.com/tasmota/arduino-esp32/releases/download/2.0.4.1/framework-arduinoespressif32.tar.gz
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Error: Traceback (most recent call last):
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/__main__.py", line 102, in main
    cli()  # pylint: disable=no-value-for-parameter
  File "/root/platformio-core/lib/python3.6/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/root/platformio-core/lib/python3.6/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/cli.py", line 71, in invoke
    return super().invoke(ctx)
  File "/root/platformio-core/lib/python3.6/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/root/platformio-core/lib/python3.6/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/root/platformio-core/lib/python3.6/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/root/platformio-core/lib/python3.6/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/run/cli.py", line 156, in cli
    verbose,
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/run/cli.py", line 201, in process_env
    result = {"env": name, "duration": time(), "succeeded": ep.process()}
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/run/processor.py", line 87, in process
    "piotest_running_name": build_vars.get("piotest_running_name"),
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/package/commands/install.py", line 131, in install_project_env_dependencies
    _install_project_env_platform(project_env, options),
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/package/commands/install.py", line 153, in _install_project_env_platform
    force=options.get("force"),
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/package/manager/platform.py", line 62, in install
    p.install_required_packages(force=force)
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/platform/_packages.py", line 76, in install_required_packages
    self.install_package(name, force=force)
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/platform/_packages.py", line 70, in install_package
    return self.pm.install(spec or self.get_package_spec(name), force=force)
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/package/manager/_install.py", line 48, in install
    pkg = self._install(spec, skip_dependencies=skip_dependencies, force=force)
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/package/manager/_install.py", line 98, in _install
    pkg = self.install_from_uri(spec.uri, spec)
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/package/manager/_install.py", line 201, in install_from_uri
    return self._install_tmp_pkg(pkg_item)
  File "/root/platformio-core/lib/python3.6/site-packages/platformio/package/manager/_install.py", line 293, in _install_tmp_pkg
    shutil.copytree(tmp_pkg.path, dst_pkg.path, symlinks=True)
  File "/root/platformio-core/lib64/python3.6/shutil.py", line 365, in copytree
    raise Error(errors)
shutil.Error: [('/root/.platformio/.cache/tmp/pkg-installing-3she5lb5/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libesp_hw_support.a', '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libesp_hw_support.a', '[Errno 28] No space left on device'), ('/root/.platformio/.cache/tmp/pkg-installing-3she5lb5/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libesp_system.a', '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libesp_system.a', "[Errno 28] No space left on device: '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libesp_system.a'"), ('/root/.platformio/.cache/tmp/pkg-installing-3she5lb5/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libfreertos.a', '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libfreertos.a', "[Errno 28] No space left on device: '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libfreertos.a'"), ('/root/.platformio/.cache/tmp/pkg-installing-3she5lb5/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libspi_flash.a', '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libspi_flash.a', "[Errno 28] No space left on device: '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/libspi_flash.a'"), ('/root/.platformio/.cache/tmp/pkg-installing-3she5lb5/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/sections.ld', '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/sections.ld', "[Errno 28] No space left on device: '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/qio_qspi/sections.ld'"), ('/root/.platformio/.cache/tmp/pkg-installing-3she5lb5/framework-arduinoespressif32/tools/sdk/esp32s3/sdkconfig', '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/sdkconfig', "[Errno 28] No space left on device: '/root/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/sdkconfig'"), ('/root/.platformio/.cache/tmp/pkg-installing-3she5lb5/framework-arduinoespressif32/variants', '/root/.platformio/packages/framework-arduinoespressif32/variants', "[Errno 28] No space left on device: '/root/.platformio/packages/framework-arduinoespressif32/variants'"), ('/root/.platformio/.cache/tmp/pkg-installing-3she5lb5/framework-arduinoespressif32/.piopm', '/root/.platformio/packages/framework-arduinoespressif32/.piopm', "[Errno 28] No space left on device: '/root/.platformio/packages/framework-arduinoespressif32/.piopm'")]

============================================================

An unexpected error occurred. Further steps:

* Verify that you have the latest version of PlatformIO using
  `pip install -U platformio` command

* Try to find answer in FAQ Troubleshooting section
  https://docs.platformio.org/page/faq/index.html

* Report this problem to the developers
  https://github.com/platformio/platformio-core/issues

============================================================
'''

I then tried the suggestion above;

'''
# pip install -U platformio
Requirement already satisfied: platformio in /root/platformio-core/lib/python3.6/site-packages (6.1.4)
Requirement already satisfied: ajsonrpc==1.* in /root/platformio-core/lib/python3.6/site-packages (from platformio) (1.2.0)
Requirement already satisfied: uvicorn==0.16.0 in /root/platformio-core/lib/python3.6/site-packages (from platformio) (0.16.0)
Requirement already satisfied: starlette==0.19.1 in /root/platformio-core/lib/python3.6/site-packages (from platformio) (0.19.1)
Requirement already satisfied: pyelftools<1,>=0.27 in /root/platformio-core/lib/python3.6/site-packages (from platformio) (0.29)
Requirement already satisfied: colorama in /root/platformio-core/lib/python3.6/site-packages (from platformio) (0.4.5)
Requirement already satisfied: bottle==0.12.* in /root/platformio-core/lib/python3.6/site-packages (from platformio) (0.12.23)
Requirement already satisfied: tabulate==0.8.* in /root/platformio-core/lib/python3.6/site-packages (from platformio) (0.8.10)
Requirement already satisfied: pyserial==3.5.* in /root/platformio-core/lib/python3.6/site-packages (from platformio) (3.5)
Requirement already satisfied: click==8.0.4 in /root/platformio-core/lib/python3.6/site-packages (from platformio) (8.0.4)
Requirement already satisfied: marshmallow==3.14.1 in /root/platformio-core/lib/python3.6/site-packages (from platformio) (3.14.1)
Requirement already satisfied: requests==2.27.1 in /root/platformio-core/lib/python3.6/site-packages (from platformio) (2.27.1)
Requirement already satisfied: zeroconf<1 in /root/platformio-core/lib/python3.6/site-packages (from platformio) (0.38.4)
Requirement already satisfied: wsproto==1.0.0 in /root/platformio-core/lib/python3.6/site-packages (from platformio) (1.0.0)
Requirement already satisfied: semantic-version==2.10.* in /root/platformio-core/lib/python3.6/site-packages (from platformio) (2.10.0)
Requirement already satisfied: aiofiles==0.8.* in /root/platformio-core/lib/python3.6/site-packages (from platformio) (0.8.0)
Requirement already satisfied: importlib-metadata in /root/platformio-core/lib/python3.6/site-packages (from click==8.0.4->platformio) (4.8.3)
Requirement already satisfied: certifi>=2017.4.17 in /root/platformio-core/lib/python3.6/site-packages (from requests==2.27.1->platformio) (2022.6.15)
Requirement already satisfied: idna<4,>=2.5 in /root/platformio-core/lib/python3.6/site-packages (from requests==2.27.1->platformio) (3.3)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in /root/platformio-core/lib/python3.6/site-packages (from requests==2.27.1->platformio) (1.26.12)
Requirement already satisfied: charset-normalizer~=2.0.0 in /root/platformio-core/lib/python3.6/site-packages (from requests==2.27.1->platformio) (2.0.12)
Requirement already satisfied: contextlib2>=21.6.0 in /root/platformio-core/lib/python3.6/site-packages (from starlette==0.19.1->platformio) (21.6.0)
Requirement already satisfied: typing-extensions>=3.10.0 in /root/platformio-core/lib/python3.6/site-packages (from starlette==0.19.1->platformio) (4.1.1)
Requirement already satisfied: anyio<5,>=3.4.0 in /root/platformio-core/lib/python3.6/site-packages (from starlette==0.19.1->platformio) (3.6.1)
Requirement already satisfied: asgiref>=3.4.0 in /root/platformio-core/lib/python3.6/site-packages (from uvicorn==0.16.0->platformio) (3.4.1)
Requirement already satisfied: h11>=0.8 in /root/platformio-core/lib/python3.6/site-packages (from uvicorn==0.16.0->platformio) (0.13.0)
Requirement already satisfied: dataclasses in /root/platformio-core/lib/python3.6/site-packages (from wsproto==1.0.0->platformio) (0.8)
Requirement already satisfied: ifaddr>=0.1.7 in /root/platformio-core/lib/python3.6/site-packages (from zeroconf<1->platformio) (0.2.0)
Requirement already satisfied: sniffio>=1.1 in /root/platformio-core/lib/python3.6/site-packages (from anyio<5,>=3.4.0->starlette==0.19.1->platformio) (1.2.0)
Requirement already satisfied: contextvars in /root/platformio-core/lib/python3.6/site-packages (from anyio<5,>=3.4.0->starlette==0.19.1->platformio) (2.4)
Requirement already satisfied: zipp>=0.5 in /root/platformio-core/lib/python3.6/site-packages (from importlib-metadata->click==8.0.4->platformio) (3.6.0)
Requirement already satisfied: immutables>=0.9 in /root/platformio-core/lib/python3.6/site-packages (from contextvars->anyio<5,>=3.4.0->starlette==0.19.1->platformio) (0.18)
'''

**Comment by dac2020** [comment]
Ah, Centos might be out of space. Making more space and trying again.

**Reply by barbudor** [reply]
Before trying to change things in user_config_override, you should be sure that you can compile without any changes
If you have installed PlatformIO manually and not using Visual Studio Code, the command line to compile `tasmota` for ESP8266 is
```
platformio run -e tasmota
```
For ESP32 it would be
```
platformio run -e tasmota32
```

You are mistaking normal ticks '  with backticks that you can insert using the `<>` icon on the tool bar right above where you are typing your text

**Reply by barbudor** [reply]
Considering that the installation of components has failed in the middle and left things in a bad state, I would recommend to `rm -Rf` all `platformio*` and `.platformio` (hidden folder) folders in `/root/` before re-trying

It also happens that PlatformIO mess-up those folders by itself and clearing them helps sorting out such sittuation.

**Reply by sfromis** [reply]
What I've found to work well for me has been to use the local build setup with [PlatformIO running in Docker](https://github.com/tasmota/docker-tasmota), this removes many risk factors from how to get PlatformIO to blend nicely into your own operating system setup.

**Reply by dac2020** [reply]
I just followed this doc.
https://tasmota.github.io/docs/Create-your-own-Firmware-Build-without-IDE/#configure-the-sources

Not worried that it didn't work the first time. I typically would run something like this to get to a point where it simply runs, then I would add my own config. I didn't change much in the config, just the ssid and pass so the device could connect.

Yes, I deleted everything and started over in another directory owned by a user, non root.
This time, it didn't error out but it just ended, still sitting there at the same spot this morning.

```
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [=====     ]  50.6% (used 41464 bytes from 81920 bytes)
Flash: [======    ]  62.2% (used 637320 bytes from 1023984 bytes)
Building .pio/build/tasmota-VN/firmware.bin
Creating BIN file ".pio/build/tasmota-VN/firmware.bin" using "/root/.platformio/packages/framework-arduinoespressif8266/bootloaders/eboot/eboot.elf" and ".pio/build/tasmota-VN/firmware.elf"
bin_map_copy([".pio/build/tasmota-VN/firmware.bin"], [".pio/build/tasmota-VN/firmware.elf"])
map_gzip([".pio/build/tasmota-VN/firmware.bin"], [".pio/build/tasmota-VN/firmware.elf"])
bin_gzip([".pio/build/tasmota-VN/firmware.bin"], [".pio/build/tasmota-VN/firmware.elf"])
Compression reduced firmware size by 71% (was 641472 bytes, now 457362 bytes)
================================================================================================= [SUCCESS] Took 84.44 seconds =================================================================================================

Processing tasmota32_base (framework: arduino; platform: https://github.com/tasmota/platform-espressif32/releases/download/v2.0.4.1/platform-espressif32-2.0.4.1.zip; board: esp32_4M)
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Tool Manager: Installing https://github.com/tasmota/arduino-esp32/releases/download/2.0.4.1/framework-arduinoespressif32.tar.gz
Unpacking  [####################################]  100%
Tool Manager: framework-arduinoespressif32@2.0.4 has been installed!
Tool Manager: Installing espressif/toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch3
Downloading  [------------------------------------]    0%
```
I'll try again with the suggestions above. 
I've never used Docker so that would be yet another thing to have to learn in a short amount of time. I'll look.

**Reply by Jason2866** [reply]
I see nothing wrong. If you do not specify a build. Every variant will be builded. Read what Barbudor wrote.
`platformio run -e tasmota32`

**Reply by sfromis** [reply]
If it is "sitting" at the spot of downloading the toolchain, you may have internet connectivity trouble, somehow

**Comment by dac2020** [comment]
I have the os ready to build the docker container but I'd like to complete this first.
As mentioned, yes, I guess I built everything but at least it completed :).

Now, I need to find the correct build name and what ever other options I need at the command line to build for these for example?
https://templates.blakadder.com/avatar_AWP12L.html

**Reply by barbudor** [reply]
No need to build the docker, just pull it from Docker's hub : `docker pull blakadder/docker-tasmota:latest`
Or even simpler, let the build script do it for you

for the listed device, it is an ESP8266, so you just need to build `tasmota`: do `./compile.sh tasmota`

**Reply by dac2020** [reply]
I see. Using openwrt for example, you would enter the device name they use and it would build that version. I'll give this a try.

**Reply by dac2020** [reply]
Ok, so I end up with;
-rwxr-xr-x  1 root root 625K Aug 30 12:59 tasmota.bin
I don't have any method of recovery so here goes nothing :)

**Reply by dac2020** [reply]
AWP12L
fanswitch
Upload Failed
Not enough space

Strange, it's 625K and the device has 1M.

**Reply by barbudor** [reply]
The device doesn't have 1M **free**
The new firmware needs to fit into the remaining part
This is the purpose of the intermediate `tasmota-minimal` firmware: a stripped down version of tasmota without any "usefull" feature but firmware update.
When using the OTA by local file upload (the 2nd part of the "Update" page) you have to manually go through the "tasmota-minimal" intermediate firmware.
When doing OTA by http download, Tasmota knows that and do that process in the background.

So:
1 - Upgrade your device with "tasmota-minimal.bin.gz" (download it from https://ota.tasmota.com/tasmota)
2 - After restart upgrade from minimal to your newly built firmware. Normally you should have also that `tasmota.bin.gz` which increase the chance of succesfully upgrading

**Reply by dac2020** [reply]
Ok, that makes sense. So I went back to minimal then upgraded using my generated version.

I still don't see a TLS (MQTT TLS) option.

**Reply by sfromis** [reply]
Then you probably did not do the required steps to add TLS support to your build.

**Reply by dac2020** [reply]
I thought someone said in the other post I had that it's there by default when I build my own so no, I didn't add it.

**Reply by sfromis** [reply]
If someone said that, someone said something wrong. While inclusion of TLS support by now is default in ESP32 builds, not for ESP8266.
https://tasmota.github.io/docs/TLS/#compiling-tls-for-esp8266

**Reply by dac2020** [reply]
Maybe that's what was said. 
Anyhow, ok then, off to look for info on how to add TLS and rebuild.

This seems to be the info. Fingerprint stuff making me quite nervous. Bit over my head.
https://tasmota.github.io/docs/TLS/#limitations

**Reply by dac2020** [reply]
LOL...
Based on the doc above, I do have lib.ssl and added the additional information in the config file.
I rebuilt using TLS and get a file below 1M.
-rwxr-xr-x  1 root   root 681K Aug 30 16:09 tasmota.bin

MINIMAL firmware
please upgrade
AWP12L
Tasmota
Upload Failed
Not enough space

I hope this doesn't mean having to go to a full code compiler. I assume there are options to also remove packages when compiling so I guess I need to dig up which packages I don't need.

**Reply by sfromis** [reply]
For OTA updates, do not use `tasmota.bin` but instead the compressed `tasmota.bin.gz` file. If you update by file upload, you need to do it in two-steps, first `tasmota-minimal.bin.gz` and then your `tasmota.bin.gz`.

**Reply by dac2020** [reply]
I have to compress it first, ok, I'll try that. I didn't read that anywhere so wasn't aware.

**Reply by dac2020** [reply]
Nice TLS is there now. Now I have to read about all the warnings of fingerprints.

**Reply by dac2020** [reply]
Works perfectly. It is now connecting to my TLS broker and communicating as it hsould.

Thank you very much again.

**Reply by sfromis** [reply]
No, you do not have to compress it first. The build process already builds the compressed .gz file for you, along the uncompressed one.

**Reply by dac2020** [reply]
Well, then I don't what changed. All I did was to compress both files and re-ran the updated. 
I first ran the gzipped minimal to get to minimal, then uploaded mine, gzipped this time and all worked.

TLS is working to some extent at least since once I enabled it, it started communicating with the broker.

**Reply by dac2020** [reply]
I saw as mentioned that there is a web based compiler too but have not installed that yet. Have to find some info on installing it on centos.

**Reply by sfromis** [reply]
If you use one of the web based build setups, you do not need to install *anything* on your operating system, except for having a decent web browser available.

**Reply by barbudor** [reply]
> I saw as mentioned that there is a web based compiler too but have not installed that yet. Have to find some info on installing it on centos.

The good thing with web based is that you have nothing to install 😄 
There are indeed 2 types of online solution:
- TasmoCompiler: https://github.com/benzino77/tasmocompiler
   This is simplified compile workflow
   You just have to click boxes, makes some selection 
   Then click the compile button 
   Not meant to develop, just compile what is existing with options

- Gitpod : This is Visual Studio Code on line.
   You can create a workspace in Gitpod, clone the repository (better to fork it before) and then do full dev in the browser, including all git operations, PR, etc ..
   The virtual machibe used by Gitpod is quite fast and I often use that because my personnal laptop is quite old

But now if you have the docker-compiler install, you are probably good to go until you want to do development

**Reply by dac2020** [reply]
I did look at that and wasn't sure what needed to be done. You said nothing to install but it sounds like there's something to install LOL. I'll look again.

I think mainly, I'll just need to add/remove some packages based on space.
It would be interesting to build some basic things like being able to traceroute a location from the device, basic TCP/IP functions.
